railties/lib/rails/generators/rails/app/templates/config/puma.rb.tt

```diff
@@ -2,38 +2,57 @@
 # are invoked here are part of Puma's configuration DSL. For more information
 # about methods provided by the DSL, see https://puma.io/puma/Puma/DSL.html.
 
-# Puma can serve each request in a thread from an internal thread pool.
-# The `threads` method setting takes two numbers: a minimum and maximum.
-# Any libraries that use thread pools should be configured to match
-# the maximum value specified for Puma. Default is set to 5 threads for minimum
-# and maximum; this matches the default thread size of Active Record.
-max_threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }
-min_threads_count = ENV.fetch("RAILS_MIN_THREADS") { max_threads_count }
-threads min_threads_count, max_threads_count
-
-# Specifies that the worker count should equal the number of processors in production.
-if ENV["RAILS_ENV"] == "production"
-  require "concurrent-ruby"
-  worker_count = Integer(ENV.fetch("WEB_CONCURRENCY") { Concurrent.physical_processor_count })
-  workers worker_count if worker_count > 1
-end
+rails_env = ENV.fetch("RAILS_ENV", "development")
+
+# Puma starts a configurable number of processes (workers) and each process
+# serves each request in a thread from an internal thread pool.
+#
+# The ideal number of threads per worker depends both on how much time the
+# application spends waiting for IO operations and on how much you wish to
+# to prioritize throughput over latency.
+#
+# As a rule of thumb, increasing the number of threads will increase how much
+# traffic a given process can handle (throughput), but due to CRuby's
+# Global VM Lock (GVL) it has diminishing returns and will degrade the
+# response time (latency) of the application.
+#
+# The default is set to 3 threads as it's deemed a decent compromise between
+# throughput and latency for the average Rails application.
+#
+# Any libraries that use a connection pool or another resource pool should
+# be configured to provide at least as many connections as the number of
+# threads. This includes Active Record's `pool` parameter in `database.yml`.
+threads_count = ENV.fetch("RAILS_MAX_THREADS") { 3 }
+threads threads_count, threads_count
 
-# Specifies the `worker_timeout` threshold that Puma will use to wait before
-# terminating a worker in development environments.
-worker_timeout 3600 if ENV.fetch("RAILS_ENV", "development") == "development"
+if rails_env == "production"
+  # If you are running more than 1 thread per process, the workers count
+  # should be equal to the number of processors (CPU cores) in production.
+  #
+  # It defaults to 1 because it's impossible to reliably detect how many
+  # CPU cores are available. Make sure to set the `WEB_CONCURRENCY` environment
+  # variable to match the number of processors.
+  processors_count = Integer(ENV.fetch("WEB_CONCURRENCY") { 1 })
+  if processors_count > 1
+    workers worker_count
+  else
+    preload_app!
```

Comment:
Just curious, why preloading the app if processors count is not > 1?

The previous given instructions were
>\# Use the `preload_app!` method when specifying a `workers` number.
>\# This directive tells Puma to first boot the application and load code
>\# before forking the application. This takes advantage of Copy On Write
>\# process behavior so workers use less memory.

Reply:
`preload_app!` is always preferable. The I don't set it when `processor_count > 1` is because it's the default there: https://www.rubydoc.info/gems/puma/Puma%2FDSL:preload_app!

> Preload the application before starting the workers; this conflicts with phased restart feature. On by default if your app uses more than 1 worker.

Reply:
That won't be fully true until Puma v7 https://github.com/puma/puma/issues/3044

Reply:
Hum, so with the current config `preload_app!` is needed regardless of the number of workers?

Reply:
https://github.com/puma/puma/blob/master/5.0-Upgrade.md#upgrade
(without looking at the code) I think its only if WEB_CONCURRENCY is used

Reply:
https://github.com/rails/rails/commit/7c1db8db9ea62f8e4b66bbeb1691ecbead80fdfe

Reply:
I was checking the history a bit to understand how this file evolved in the last time and I found the following:

1. 471ab2347fc8941ead6d9f95ef32c6686d706924 - Remove `workers` and `preload_app!`. Let Puma take care of the defaults
2. 839ac1e70f7fd340a93348d54cde626fbdbedcb6 - Try to use all CPU cores by default
3. 06d614ada9e4609ff83659e842f48af3232a03a5 - Go back to a default of 1 for WEB_CONCURRENCY

So maybe all of this is unnecessary and we should go back to 1. ?